### PR TITLE
Add support for lazy vector generation to expression fuzzer

### PR DESCRIPTION
**Pull request by bikramSingh91**
This patch leverages the ability to wrap a vector in lazy encoding
that was recently added in the VectorFuzzer and enables it to run
the same input in both the common and simplified evaluation path
but with the common path having some inputs wrapped in lazy
encoding. The columns to be wrapped are randomly chosen with a
probability that can be specified by a gflag
"lazy_vector_generation_ratio".

Additionally
- context about which columns were chosen to be wrapped are also
seriazlized with other data required to repro in case of an error.
- The repro data is now stored in a newly generated folder with a
unique name.

### CODE REVIEW

File: velox/vector/VectorSaver.cpp [hunk 1]
```diff
@@ -678,4 +678,31 @@ std::optional<std::string> generateFilePath(
   }
   return path;
 }
+
+std::optional<std::string> generateFolderPath(
```
Comment (tanjialiang) [inline velox/vector/VectorSaver.cpp:682]: Is this going to generate path recursively if middle directories do not exist?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 2]
```diff
@@ -17,16 +17,34 @@
...
 namespace facebook::velox::test {
 
 namespace {
 
-void printRowVector(const RowVectorPtr& rowVector) {
-  LOG(INFO) << "RowVector contents (" << rowVector->type()->toString() << "):";
+// File names used to persist data required for reproducing a failed test case.
+static const char* kInputVectorFileName = "inputVector";
+static const char* kIndicesOfLazyColumnsFileName = "indicesOfLazyColumns";
+static const char* kResultVectorFileName = "resultVector";
```
Comment (tanjialiang) [inline velox/expression/tests/ExpressionVerifier.cpp:29]: Maybe change file names to input_vector, result_vector, indices_of_lazy_columns to conform with the convention

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 3]
```diff
@@ -18,16 +18,34 @@
...
+#include "velox/vector/fuzzer/VectorFuzzer.h"
 
 namespace facebook::velox::test {
 
 namespace {
 
-void printRowVector(const RowVectorPtr& rowVector) {
-  LOG(INFO) << "RowVector contents (" << rowVector->type()->toString() << "):";
+// File names used to persist data required for reproducing a failed test case.
+static const char* kInputVectorFileName = "inputVector";
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:28]: the recommendation is to create static strings as std::string_view (so it provides a full std::string-like API). This should be described in our coding style doc.

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 4]
```diff
@@ -121,25 +139,47 @@ void compareVectors(const VectorPtr& left, const VectorPtr& right) {
   LOG(INFO) << "All results match.";
 }
 
+// Returns a copy of 'rowVector' but with the columns having indices listed in
+// 'columnsToWarpInLazy' wrapped in lazy encoding. 'columnsToWarpInLazy' should
+// be a sorted list of column indices.
+RowVectorPtr wrapColumnsInLazy(
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:145]: Can we somehow reuse `VectorFuzzer::fuzzRowChildrenToLazy()` here? (say we push the coin toss logic down to VectorFuzzer).

Comment (bikramSingh91) [reply velox/expression/tests/ExpressionVerifier.cpp:145]: fuzzRowChildrenToLazy() randomly selects columns to fuzz with lazy, however in this case we need to retain the information about which columns would be fuzzed so that we can recreate the same set of columns for when the test re-runs after applying TRY() (when in the first try both common and simplified path return the same error). Moreover, this also helps serialize the list for quick repro using vector fuzzer

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 5]
```diff
@@ -460,7 +464,7 @@ void ExpressionFuzzer::appendConjunctSignatures() {
 }
 
 RowVectorPtr ExpressionFuzzer::generateRowVector() {
-  return vectorFuzzer_.fuzzRow(
+  return vectorFuzzer_.fuzzInputRow(
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionFuzzer.cpp:467]: I know we have this assumptions that top-level row vectors won't have nulls, but is this breaking something? It also looks a bit orthogonal to this PR.

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 6]
```diff
@@ -121,25 +139,47 @@ void compareVectors(const VectorPtr& left, const VectorPtr& right) {
...
+    RowVectorPtr rowVector,
+    const std::vector<column_index_t>& columnsToWarpInLazy) {
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:147]: typo

File: velox/expression/tests/ExpressionVerifier.cpp (hunk 4 above)
Comment (pedroerp) [reply velox/expression/tests/ExpressionVerifier.cpp:145]: > however in this case we need to retain the information about which columns would be fuzzed

could we achieve the same by calling the VectorFuzzer method above, then iterating over the generated columns collecting which ones have `isLazy()` set?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 7]
```diff
@@ -232,64 +278,74 @@ bool ExpressionVerifier::verify(
 
 void ExpressionVerifier::persistReproInfo(
     const VectorPtr& inputVector,
+    std::vector<column_index_t> columnsToWarpInLazy,
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:281]: do we need to push down this vector here? Couldn't we infer it based on which column in inputVector have `isLazy()` set?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 8]
```diff
@@ -172,7 +212,13 @@ bool ExpressionVerifier::verify(
   try {
     exec::ExprSet exprSetCommon(
         {plan}, execCtx_, !options_.disableConstantFolding);
-    exec::EvalCtx evalCtxCommon(execCtx_, &exprSetCommon, rowVector.get());
+    auto wrappedRowVector = wrapColumnsInLazy(rowVector, columnsToWarpInLazy);
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:215]: maybe encapsulate the debuging logic in the wrapColumnsInLazy() function so that you could keep this part simpler:

```
exec::ExprSet exprSetCommon(
        {plan}, execCtx_, !options_.disableConstantFolding);
exec::EvalCtx evalCtxCommon(
        execCtx_, &exprSetCommon, wrapColumnsInLazy(rowVector));
```

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 5 above)
Comment (bikramSingh91) [reply velox/expression/tests/ExpressionFuzzer.cpp:467]: Currently the way the velox works, we only have lazyVectors as children in row vectors is when its an input vector (no null children). Ideally expressions are also usually called on input vectors so I thought this would be ok.
I have considered a few arguments for an against this change, let me know what you think:

For (using fuzzInputRow):
- More closely match how we usually run expressions
- columns (to wrap in lazy) can be randomly selected without worrying about nullability. Otherwise, having null columns would result in not hitting the probability of selecting lazy columns.

Against (using fuzzRow):
- We gain coverage on running the fuzzed expression on a vector that can represent a column of row type with possible null children. Not sure how much coverage we will get here vs using fuzzInputRow() which can still have children that are rows.

I dont have a strong preference either way as the arguments are more about justifying randomness which is a hard thing to prove or compare in this instance.

File: velox/expression/tests/ExpressionVerifier.cpp (hunk 4 above)
Comment (bikramSingh91) [reply velox/expression/tests/ExpressionVerifier.cpp:145]: That was my first attempt too, but the problem then arises is when we need to retain that information when we re-run the same expression but with TRY() added (https://github.com/facebookincubator/velox/blob/043e397822cdbf38b42717634fec127879e42d39/velox/expression/tests/ExpressionFuzzer.cpp#L730-L743), this is where we need to retain the information about which columns were fuzzed and have a utility function wrapColumnsInLazy() to apply it. If I instead use fuzzRowChildrenToLazy(), I would have to add that code in ExpressionFuzzer.cpp that extracts the info about fuzzer column idxs, then pass 2 different row vectors (Since lazy vectors cannot be copied, I'll need a utility to apply the same wrapping again when re-running with TRY). Basically the amount of code would be the same, just some of it would moved around between ExpressionFuzzer and ExpressionVerifier

Comment (pedroerp) [reply velox/expression/tests/ExpressionVerifier.cpp:156]: I see, it makes more sense now. Thanks for explaining it!

File: velox/expression/tests/ExpressionFuzzer.cpp [hunk 9]
```diff
@@ -709,14 +713,25 @@ void ExpressionFuzzer::go() {
       resultVector = vectorFuzzer_.fuzzFlat(plan->type());
     }
 
+    // Randomly pick columns from the input row vector to wrap in lazy.
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionFuzzer.cpp:716]: nit: wrapping these things into smaller helper function tend to make the code more readable:

`auto columnToWrapInLazy = generateLazyColumnIds();`

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 10]
```diff
@@ -131,25 +150,50 @@ void compareVectors(const VectorPtr& left, const VectorPtr& right) {
   LOG(INFO) << "All results match.";
 }
 
+// Returns a copy of 'rowVector' but with the columns having indices listed in
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:153]: Should we actually provide this function as part of VectorFuzzer? Maybe an overload of `VectorFuzzer::wrapInLazyVector()` that takes a vector of column ids?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 11]
```diff
@@ -131,25 +150,50 @@ void compareVectors(const VectorPtr& left, const VectorPtr& right) {
...
+        i == columnsToWrapInLazy[listIndex]) {
+      listIndex++;
+      child = VectorFuzzer::wrapInLazyVector(child);
+    }
+    children.push_back(child);
+  }
+  auto lazyRowVector = std::make_shared<RowVector>(
+      rowVector->pool(),
+      rowVector->type(),
+      nullptr,
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:178]: should we pass the same null buffer, just in case?

File: velox/expression/tests/ExpressionFuzzer.cpp (hunk 5 above)
Comment (bikramSingh91) [reply velox/expression/tests/ExpressionFuzzer.cpp:459]: No, it would still work, just saying that if we use "RowVectors with null top-level" we wont hit the targeted proportion of columns that we want to wrap in lazy ().

File: velox/expression/tests/ExpressionVerifier.cpp (hunk 10 above)
Comment (bikramSingh91) [reply velox/expression/tests/ExpressionVerifier.cpp:153]: I was thinking the same, but we probably dont want to overload the VectorFuzzer with too many options just yet, if there are similar use cases in other places in the future, we can consolidate those and then add this functionality to the fuzzer. What do you think?

File: velox/expression/tests/ExpressionVerifier.cpp [hunk 12]
```diff
@@ -131,25 +150,61 @@ void compareVectors(const VectorPtr& left, const VectorPtr& right) {
...
+      child = VectorFuzzer::wrapInLazyVector(child);
+    }
+    children.push_back(child);
+  }
+
+  BufferPtr newNulls = nullptr;
+  if (rowVector->nulls()) {
+    newNulls = AlignedBuffer::allocate<bool>(
+        rowVector->nulls()->size(), rowVector->pool());
+    memcpy(
```
Comment (pedroerp) [inline velox/expression/tests/ExpressionVerifier.cpp:180]: use buffer->copy() or copyFrom API so you don't need to memcpy explicitly?

**Review by pedroerp — APPROVED** [review]
Looks good to me. Just a small comment (plus also double check @tanjialiang 's comment) .
